cfsqlite-triggers.c

```diff
@@ -1,10 +1,80 @@
 #include "cfsqlite-triggers.h"
 #include "cfsqlite-tableinfo.h"
+#include "cfsqlite-util.h"
 
 #include <stdint.h>
 #include <string.h>
+#include <stdio.h>
 
-int cfsql_createCrrViewTriggers(
+char *cfsql_conflictSetsStr(cfsql_ColumnInfo *cols, int len)
+{
+  // set statements...
+  char *sets[len];
+  int resultLen = 0;
+  char *ret = 0;
+
+  for (int i = 0; i < len; ++i)
+  {
+    if (cols[i].versionOf != 0)
+    {
+      sets[i] = sqlite3_mprintf(
+        "\"%s\" = CASE WHEN EXCLUDED.\"%s\" THEN \"%s\" + 1 ELSE \"%s\" END",
+        cols[i].name,
+        cols[i].versionOf,
+        cols[i].name,
+        cols[i].name
+      );
+    }
+    else
+    {
+      sets[i] = sqlite3_mprintf("\"%s\" = EXCLUDED.\"%s\"", cols[i].name, cols[i].name);
+    }
+
+    resultLen += strlen(sets[i]);
+  }
+  resultLen += len - 1;
+  ret = sqlite3_malloc(resultLen * sizeof(char) + 1);
+  ret[resultLen] = '\0';
+
+  cfsql_joinWith(ret, sets, len, ',');
+  return ret;
+}
+
+char *cfsql_localInsertOnConflictStr(cfsql_TableInfo *tableInfo)
+{
+  if (tableInfo->pksLen == 0)
+  {
+    // dup given the caller would try to deallocate it and we
+    // cannot deallocate a literal
+    return strdup("");
+  }
+
+  char *pkList = cfsql_asIdentifierList(tableInfo->pks, tableInfo->pksLen, 0);
+  char *conflictSets = cfsql_conflictSetsStr(tableInfo->nonPks, tableInfo->nonPksLen);
+
+  char *ret = sqlite3_mprintf(
+      "ON CONFLICT (%s) DO UPDATE SET\
+      %s%s\
+    \"__cfsql_cl\" = CASE WHEN \"__cfsql_cl\" %% 2 = 0 THEN \"__cfsql_cl\" + 1 ELSE \"__cfsql_cl\" END,\
```

Comment:
would this indiscriminate incrementing of the `cl` break if a client is doing `upserts` rather than updates?